### ISSUE TITLE
Cubic Bézier Component

### ISSUE THREAD

**Issue by ksassnowski** [motion-canvas/motion-canvas#312]
**Description**

There's already a `Line` component to draw polylines, so I think having a `CubicBezier` component would be a useful next step (rather than shooting straight for splines). Since Canvas already has a `bezierCurveTo` method to draw cubic Béziers, most of the work is already done for us, anyways.

**Proposed solution**

I have implemented a quick proof of concept for a Bézier curve in one of my own projects which looks something like this when rendered:

![image](https://user-images.githubusercontent.com/5139098/218241829-5097111e-9bbe-4c1a-810e-747c1588bf34.png)

The API currently looks like this

```tsx
 <CubicBezier
    lineWidth={6}
    stroke={"white"}
    from={[320, -200]}
    fromHandle={[0, -200]}
    toHandle={[-400, 200]}
    to={[-600, 200]}
/>
```

The exact names for the points is obviously up for debate. There should probably also a way for creating a curve from just one list of numbers like `[p0x, p0y, cp1x, cp1y, cp2x, cp2y, p1x, p2y]` 

```tsx
 <CubicBezier
    stroke={"white"}
    points={[320, -200, 0, -200, -400, 200, -600, 200]}
/>
```

or as a list of four `PossibleVector2`.

```tsx
 <CubicBezier
    stroke={"white"}
    points={[[320, -200], [0, -200], [-400, 200], [-600, 200]]}
/>
```

I'd propose drawing the overlay for the component like I have done in my screenshot. Visually distinguishing the start and end points from the control points and drawing the connecting lines between is useful when trying to get the shape of the curve right.

As far as I can tell, it wouldn't make much sense for this component to extend `Line` since the `Line` component depends pretty heavily on its `point` array. Having a distinct set of `points` doesn't really make sense for a continuous curve like that. So I propose for it to extend from `Shape`, instead.

The biggest challenge I've found so far is animating the curve similar to how a `Line`'s `start` and `end` parameters work. Drawing a partial Bézier curve is possible, but fairly awkward. It seems to always revolve around calculating the length of the curve and then setting the `lineDash` and `lineDashOffset` properties so that only a part of the curve is rendered. I've found this StackOverflow answer that seems to do just that: https://stackoverflow.com/a/69789529

I've also implemented a `getPointAtPercentage` method that works similar to the `Line` component's method  (sans the tangent at that point, but that could be added as well). This would when trying to animate something along the curve. This can be done fairly simply using the polynomial form of the curve, for example.

**Considered alternatives**

I'd be happy to take a crack at this but will probably need some guidance on what other properties the component should expose as well. Looking at the `Line` component, there's also a public `profile` method that returns a `CurveProfile` for the line. I'm not entirely sure where this is used outside the component itself or if the `CubicBezier` component should expose something like that as well.

**Comment by gaetgu** [comment]
I came here to open this exact same issue. I am working on an animation using motion canvas, and it would be really nice—eventually essential—to have a beziér curve. I would be happy to help out in the implementation of this as I have worked on implementing cubic and quadratic beziér curves in the past, but I do not know the codebase very well yet. If you want me to do something, just let me know what and I will figure it out.

Thanks for the awesome work on this so far!

**Comment by thoughtlessnerd** [comment]
Hello there,
Can you assign me to work on creating Cubic Bezier component ??

**Comment by aarthificial** [comment]
This is a really good proposal for a single Bezier curve. 
I think we should also go for a composite Bezier curve, like in other animation software. 
Something along the lines:
```tsx
<Spline>
  <Knot endHandle={[10, 10]} />
  <Knot position={[100, 100]} startHandle={[-10, 10]} endHandle={[0, -10]} broken />
  <Knot position={[200, 50]} startHandle={[10, 0]}/>
</Spline>
```
where the line is defined not by individual Bezier segments, but rather by the control points in between.
Maybe this should be its own issue, but it would be nice to keep it in mind when implementing the singular curve so the logic can be later reused.

It's also important to reflect the Line API with properties like `start` end `end` that allow you to animate the line appearing on the screen. `startOffset` and `endOffset` that cuts out the line at a specific distance from the `start`/`end`. And `startArrow`, `endArrow`.

To whoever will implement this, these things will require some mathematical knowledge when it comes to splitting bezier curves in two, and such.

**Comment by ksassnowski** [comment]
In this example, I assume the `startHandle` and `endHandle` properties are relative to the `Knot`'s `position`?

**Comment by aarthificial** [comment]
Yes, I think relative-first is the best approach in situations like this.
Using signals you can always do this: 
```ts
knot().endHandle(() => node().absolutePosition().transformAsPoint(knot().worldToLocal()))
```
To pin the handle to some other node.

**Comment by thoughtlessnerd** [comment]
I'm confused, i found where i need to make the component but like, I've forked the repo, i made a branch for it but now how do i start developing ? can anyone guide me through it. thankyou

**Comment by aarthificial** [comment]
Have tried reading [README](https://github.com/motion-canvas/motion-canvas#developing-motion-canvas-locally)?
Feel free to join our Discord server if you have any problems with the steps described there.

**Comment by GufNZ** [comment]
From memory, splines are a superset of Beziers, as you can get a spline to hit a knot/control point by just stacking them up enough times, so once splines are implemented properly, Beziers can just be a helper function that generates the correct set of knots.
It's been a while since I was looking at them so I forget the math for splitting splines, but I'm pretty sure it is reasonably straight forward.  Something like: find the local tangent, add CPs stacked on the cut point so the curve still hits it plus a couple to keep the tangent the same, then we can use the stacked ones as both the end of the first half and the start of the second.

**Comment by aarthificial** [comment]
Closed by #514
